### DISCUSSION TITLE
Need some advice for GAIL implement

### DISCUSSION THREAD

**Discussion by nicholas0717** [Toni-SM/skrl#40] [Q&A]
Hi @Toni-SM 

I'm trying to use skrl library to implement GAIL to perform the task `FactoryTaskNutBoltPick` in Isaac Gym. But it works badly. I need some advice.

Firstly, my idea is to use `PPO`  as the generator for `GAIL` , so define the `GAIL` class inheriting the `PPO` class. And I define the model `discriminator` in `GAIL` and set related configurations. The memory_exp is the expert's demonstration which size is 300000.
![image](https://user-images.githubusercontent.com/65715742/197671510-e3a7d7c5-6466-4a37-95bd-e859d5c8a33b.png)

The policy and value model is the same as `PPO` and discriminator model  is define as follow:

![image](https://user-images.githubusercontent.com/65715742/197671164-c88dc305-ec79-4c78-9821-0972b7d8d147.png)
![image](https://user-images.githubusercontent.com/65715742/197671626-938ba1ef-25d7-46b3-a858-30ea567eb684.png)

So, the `_update()` in GAIL will include two update function. First is '_update_disc' which is used to update the discriminator. Its inputs are sampled states and actions from policy and demonstrations. After `_update_disc`, calculate `disc_reward` from discriminator and update the PPO `super()._update(timestep, timesteps)` with `disc_reward` (not using policy's initial rewards). And the process of `super()._update(timestep, timesteps)` is basically consistent with the process of _update() in PPO. Use  `disc_reward` to calculate the GAE, sample and compute the loss of policy, value and entropy(same as PPO).

![image](https://user-images.githubusercontent.com/65715742/197672166-13257cf9-1151-4158-9cfb-c8a27c556d6b.png)
![image](https://user-images.githubusercontent.com/65715742/197672539-efa075f6-bfb8-4ca9-bd9b-9fea643792ac.png)

I‘m not sure if the GAIL I implemented is correct. But I used it to train `FactoryTaskNutBoltPick` and performed badly. Maybe related parameters are wrong? I need your advice. 

![image](https://user-images.githubusercontent.com/65715742/197673553-bcf77d4c-5d52-4e92-ba31-6cf0ba06322d.png)
![image](https://user-images.githubusercontent.com/65715742/197673665-46fc88da-91ea-4cbc-b264-96d0bbcfffe1.png)

**Comment by nicholas0717** [comment]
By the way, I want to know the differences between `act` and `compute` in Shared model. I'm puzzled about their using.

**Comment by Toni-SM** [comment]
Hi  @nicholas0717 

> I hope this first iteration helps in some way. 
> Don't be afraid to reply and continue the discussion...

<hr>

I would like to recommend the following:

1. Don't inherit from another agent, instead, duplicate the code of some agent in a new folder and start from there. This will allow you to have control over all the agent code and not have to navigate between base classes to inspect any implementation.
2. It would be better to use AMP as a starting agent. This agent is also based on PPO and makes use of a discriminator to update the reward.
3. Experiment with a simpler and faster environment such as Cartpole, BallBalance, or Ant

If it's ok with you, and you can make the changes (particularly points 1 and 2) it would be great if you could share the code to try to help with it.

<hr>

Regarding the distinction  between `act` and `compute` in shared (and not shared) models:

The `compute` method is responsible for calling the artificial neural network (defined by the user) to process the input data (observations and optionally actions). The artificial neural network, generally speaking, is ultimately a deterministic function (for the same input values it returns the same output values). 

The `act` method, on the other hand, contains the implementation of the model (deterministic, Gaussian, categorical, etc.) which is nothing more than a post-processing stage of the neural network output to generate the desired output (deterministic/stochastic, discrete/continuous). It calls the `compute` method to get the output of the network as a first step.

![Screenshot from 2022-10-25 12-35-03](https://user-images.githubusercontent.com/22400377/197752439-3e195a3d-5fe3-4582-805b-ff37f7d8a166.png)

**Comment by nicholas0717** [comment]
Hi @Toni-SM 

Thank you for you reply!

After I changed the discriminator's` Linear()` units, it seemed to work, but not perfect. Maybe my related parameters still need to be tuned.
![image](https://user-images.githubusercontent.com/65715742/197992144-6debb044-4278-44c0-907c-a299612dae38.png)
![image](https://user-images.githubusercontent.com/65715742/197992265-80118585-e2b6-4e76-93f8-54d82fd9ed1e.png)

I want to get some of your experience when you need to tune parameters in different tasks. They played an important role in PPO, GAIL and AMP, especially like `rollouts`, `learning_epochs`, `mini_batches` and number of network's hidden units.

Thank you so much. You help me a lot!

**Comment by Toni-SM** [comment]
Hi @nicholas0717 

Good to hear it works.

Yes, some algorithms require hyperparameter fine-tuning to work better.
There is currently no information related to hyperparameter tuning in the library documentation.
However, integration with a hyperparameter optimizer (such as [optuna](https://optuna.org/)) is on the roadmap.
It should be easy to implement, but unfortunately, I have not had the time to do so. 
It would be nice if you can try it and share your experiences...
 
Hope to do it at some free time :)
